### ISSUE TITLE
Support -100 as ignore_index of Accuracy

### ISSUE THREAD

**Issue by richarddwang** [Lightning-AI/metrics#304]
When running the instance `Accuracy(ignore_index=-100)`, it will tell you `-100` is not in the classes.

But `torch.nn.CrossEntropy(..., ignore_index=-100)` is default, and we often train masked language model with  label tensor in shape (batch_size, sequence_length) where positions not selected by masking process are filled with -100 to prevent counting loss from non-masking positions.

We can still check `ignore_index` is not > number of classes, but releasing the limit that `ignore_index` can't  < 0 would be better.

**Comment by SkafteNicki** [comment]
Hi @richarddwang, sound good to me. Would you be up for sending a PR?

The relevant check that should be changed:
https://github.com/PyTorchLightning/metrics/blob/7af6d13b3c2186aacf5594793317c15a199608bb/torchmetrics/functional/classification/stat_scores.py#L95-L96
(and a similar checks)

**Comment by richarddwang** [comment]
Hi @SkafteNicki, I am not clear about how `ignore_index` is handled,
https://github.com/PyTorchLightning/metrics/blob/7af6d13b3c2186aacf5594793317c15a199608bb/torchmetrics/functional/classification/stat_scores.py#L111-L113
https://github.com/PyTorchLightning/metrics/blob/7af6d13b3c2186aacf5594793317c15a199608bb/torchmetrics/functional/classification/stat_scores.py#L24-L27
Would it still be good when ignore_index is -100 ?

**Comment by SkafteNicki** [comment]
@richarddwang yeah, maybe it is bit more complex than first thought. Could you please provide an example input that you would normally feed into `torch.nn.CrossEntropy` with `ignore_index=-100` just so I can better understand?

**Comment by richarddwang** [comment]
Sure, one that wants to do masked language model will often see something like this.
``` python
batch_size, sequence_length  = 4,5
mlm_logits = torch.randn(4,5,128)
labels = torch.tensor([
    [1234,-100,-100,-100,-100],
    [-100,-100,-100,7567,-100]
    [-100,-100,8900,-100,-100]
    [-100,-100,-100,-100,3456]
]) # (4,5), where -100 says it is not selected position for mlm to learn from
loss = nn.CrossEntropyLoss(mlm_logits.view(-1,128), labels.view(-1))
mlm_predictions = mlm_logits.argmax(dim=-1) # (4,5)
self.log('mlm_acc', mlm_acc(mlm_predictions.view(-1), labels.view(-1)))
```

**Comment by collinmccarthy** [comment]
Can we also handle the case for segmentation with datasets like PASCAL VOC, which has a "void" label of 255 but the number of classes is 21?  I tried passing in the number of classes but then I get the error:

```
The `ignore_index` 255 is not valid for inputs with 21 classes
``` 

I can open a new issue but since this is actively being discussed and it's the same general issue I thought I'd add a comment here. Thanks!

**Comment by keshavd** [comment]
I fixed this by just masking out the ignored index.

ex.

```
import torchmetrics

ignore_index = -100
mask = ~labels.eq(ignore_index)

acc = torchmetrics.Accuracy()
acc(logits[mask], labels[mask])
```

**Comment by richarddwang** [comment]
Hi all.
I wrote a naive solution, which makes the ignored part be incorrect in comparison and then divided by number of unignored.
There may be better way, this is just for your reference.
```python
    @torch.no_grad()
    def accuracy(
        logits_or_ids,  # <float>(B,L,C)/<long>(B,L)
        labels,  # <long>(B,L)
        ignore_id: int,
    ):
        if len(logits_or_ids.shape) == len(labels.shape):
            preds = logits_or_ids  # <long>(B,L)
        else:
            preds = logits_or_ids.argmax(dim=-1)  # <long>(B,L)
        ignore = labels == ignore_id
        # make ignored positions incorrect thus out of correction counting
        assert ignore_id != -9999
        preds = preds.masked_fill(ignore, -9999)
        num_correct = (preds == labels).sum()
        num_nonignore = (~ignore).sum()
        return num_correct / num_nonignore
```

**Comment by Borda** [comment]
@SkafteNicki how are we doing here? :rabbit:

**Comment by TezRomacH** [comment]
This feature is a must-have for also `F1`, `AUCROC`, and any other classification metrics.

**Comment by alexsr** [comment]
I wrote a simple wrapper to flat out ignore the targets with `ignore_index`.
```py
from torchmetrics import functional as FM

def ignore_index(func):
    def ignore_and_call(preds: Tensor, target: Tensor, ignore_target: Optional[int] = None, *args, **kwargs):
        indices = (target != ignore_target).nonzero(as_tuple=False)
        preds_ignored = preds[indices]
        target_ignored = target[indices]
        res = func(*args, **kwargs, preds=preds_ignored, target=target_ignored)
        return res
    return ignore_and_call, func

accuracy = ignore_index(FM.accuracy)
```
